frame/oracle/design/design.adoc

```diff
@@ -0,0 +1,695 @@
+. The system MUST scale to be able to provide prices for any number of assets that Picasso DeFi ecosystem requires.
+
+=== Decentralization
+
+. The system MUST support the number of Oracle providers that governance deems necessary for the decentralization of the protocol.
+. The system MUST support a wide variety of price sources.
+
+== Method
+
+=== System Overview
+
+Following diagram illustrates an overview of the Apollo design with some minor components skipped over for clarity.
+
+.Apollo Price Oracle: The System Overview
+[plantuml,images/apollo-overview,png]
+----
+skinparam interface {
+  backgroundColor RosyBrown
+  borderColor orange
+}
+
+skinparam node {
+  backgroundColor<<parachain>> TECHNOLOGY
+}
+
+skinparam package {
+  backgroundColor LightCyan
+}
+
+skinparam cloud {
+  backgroundColor Azure
+}
+
+skinparam node {
+  backgroundColor<<price-source>> Lavender
+}
+
+skinparam databaseBackgroundColor SkyBlue
+
+cloud "Oracle Provider Infra" <<oracle-infra>> {
+    node "Picasso Parachain Node" <<parachain>> {
+        () "Oracle" as o
+        package "Pallet Apollo" {
+            Database "PrePrices" as prepricedb
+            Database "AssetsInfo" as assetdb
+            Database "OracleStake" as stakedb
+            Database "Prices" as pricedb
+            Database "PriceHistory" as pricehistorydb
+            Database "RewardTrackerStore" as rewarddb
+            [Offchain Worker] as ofw
+            [on_init hook] as oih
+            () "SubmitPrice" as sp
+            () "AdjustRewards" as ar
+            () "AddStake" as as
+            () "AddAssetAndInfo" as adda
+            ar --> rewarddb
+            ofw --> sp
+            sp --> prepricedb
+            oih <-- prepricedb
+            adda --> assetdb
+            assetdb --> oih
+            oih --> pricedb
+            oih <--> stakedb : read/slash
+            as --> stakedb
+            oih --> pricehistorydb
+        }
+
+        ["Governance"] --> ar
+        ["Governance"] --> adda
+        rewarddb --> oih : block reward
+        pricedb --> o
+        pricehistorydb --> o
+    }
+
+    package "Price Feed Server" <<price-feed>> {
+        () "Rest API" as papi
+        database "Price Cache" as pc
+        () "Client 1 - Binance Client" as c1
+        () "Client 2 - Pyth Client" as c2
+        () ".. Client n" as cn
+
+        pc --> papi
+        c1 --> pc: price feed
+        c2 --> pc: price feed
+        cn --> pc: price feed
+        c1 <.up.> binance
+        c2 <.up.> pyth
+        cn <.up.> n
+    }
+}
+
+ofw <.left.> papi : get price
+
+cloud {
+    node "Price-Source 1 - Binance" <<price-source>> as binance {
+    }
+}
+
+cloud {
+    node "Price-Source 2 - Pyth" <<price-source>> as pyth {
+    }
+}
+
+cloud {
+    node ".. Price-Source n" <<price-source>> as n {
+    }
+}
+
+"Oracle Provider" as op
+op -up-> as
+----
+
+NOTE: It is proposed to rename the existing pallet-oracle as pallet-apollo here.
```

Comment:
Can we rename vault to cubic while we're at it :laughing:

Reply:
haha sure we can, but not something relevant to this PR :)